### PR TITLE
[https://nvbugs/5615248][perf] Early emission of first token with overlap scheduling

### PR DESCRIPTION
**Pull request by brb-nv**
## Description

This MR advances `enqueue_responses()` of requests with first generated tokens before `sample_async` when using overlap scheduling. This is particularly useful for small language models with tight TTFT expectations.

It's experimental and opt-in, defaulting to false.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Summary by CodeRabbit

* **New Features**
  * Enabled early emission of first-token responses in overlap scheduling mode for improved Time-To-First-Token (TTFT) performance.
  * Enhanced streaming response behavior to maintain parity across scheduling modes.

* **Bug Fixes**
  * Prevented duplicate first-token response emission during request processing.

* **Tests**
  * Added comprehensive streaming validation tests for overlap scheduler with various configurations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14061)

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 1]
```diff
@@ -4109,7 +4156,14 @@ def _handle_responses(self):
...
             request_done = False
-            if request.py_decoding_iter == 1 or request.is_finished or \
+            # If `_emit_first_token_responses` already enqueued the
+            # first-token response earlier this iteration, skip the
+            # duplicate `create_response`. Still fall through for
+            # termination handling below.
+            if request.py_first_token_response_sent:
+                request.py_first_token_response_sent = False
+                request_done = request.is_finished
+            elif request.py_decoding_iter == 1 or request.is_finished or \
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: https://github.com/NVIDIA/TensorRT-LLM/pull/14061/commits/7c601aeab9c3d3b5c1465741acb3e63aacb5ebac

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -4050,6 +4061,42 @@ def _handle_first_token_response(self, scheduled_batch):
...
+            request.draft_tokens = request.py_draft_tokens if get_draft_token_length(
+                request) > 0 else []
+            request.decoding_iter = request.py_decoding_iter
+
+            response = request.create_response(False, self.dist.rank)
+            if response is None:
+                continue
+            response.result.cached_tokens = request.cached_tokens
+            new_responses.append((request.py_request_id, response))
+            request.py_first_token_response_sent = True
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: https://github.com/NVIDIA/TensorRT-LLM/pull/14061/commits/7c601aeab9c3d3b5c1465741acb3e63aacb5ebac

**Review by reasonsolo — APPROVED** [review]

**Review by achartier — APPROVED** [review]
LGTM

**Comment by brb-nv** [comment]
/bot run --stage-list "DGX_B200-PyTorch-4" --disable-fail-fast

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -4241,6 +4198,65 @@ def _handle_first_token_response(self, scheduled_batch):
...
+        for request in prev_scheduled_requests.all_requests():
+            if request.py_decoding_iter != 1:
+                continue
+            if request.is_attention_dp_dummy or request.is_cuda_graph_dummy:
+                continue
+            if request.py_kv_transfer_timed_out:
+                continue
+            # Disagg gen-only requests are emitted by `_handle_first_token_response`. Skip here.
+            if request.is_generation_only_request() and not request.is_finished:
+                continue
```
Comment (Tabrizian) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: non-blocking: is there a way to unify the places where we send responses back? For example, could we move handle_responses to the place where we call _emit_first_token_responses to always return responses as early as possible?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 4]
```diff
@@ -4300,7 +4316,20 @@ def _handle_responses(self):
                 request.update_perf_metrics(self.iter_counter)
 
             request_done = False
-            if request.py_decoding_iter == 1 or request.is_finished or \
+            # `_emit_first_token_responses` only enqueues non-terminal
+            # streaming responses. If the request became finished after
+            # that emit (e.g. cancelled by `_handle_canceled_requests`),
+            # issue a fresh terminal response here; otherwise skip the
+            # duplicate `create_response`.
```
Comment (Tabrizian) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: is there any way to test the condition here?

**Review by Tabrizian — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 3 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Great point, Iman. I think the current design (on main) is to call `sample_async` as soon as possible so that the GPU is kept busy and then do host work for response handling. This MR makes an exception to that to improve TTFT for small models. Advancing `_handle_responses` unconditionally for all requests could risk creating a bubble on GPU due to `sample_async` call getting delayed by a lot. So, I'm not doing that for now.

Comment (Tabrizian) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Got it, thanks for explaining Balaram.

**Comment by brb-nv** [comment]
@coderabbitai full review

File: tests/integration/test_lists/test-db/l0_h100.yml [hunk 5]
```diff
@@ -122,6 +122,7 @@ l0_h100:
   - accuracy/test_llm_api_pytorch_multimodal.py::TestGemma3_27BInstruct::test_fp8_prequantized
   - accuracy/test_llm_api_pytorch_multimodal.py::TestNanoV3Omni::test_auto_dtype[fp8]
   - accuracy/test_llm_api_pytorch.py::TestQwen3_8B::test_fp8_block_scales[latency]
+  - accuracy/test_llm_api_pytorch.py::TestQwen3_8B::test_fp8_block_scales_early_first_token_emission
```
Comment (StanleySun639) [inline tests/integration/test_lists/test-db/l0_h100.yml]: Could you help add the test case to [QA test list](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/qa/llm_function_core.txt) to gain more HW test coverage?

**Review by StanleySun639 — APPROVED** [review]

File: tests/integration/test_lists/test-db/l0_h100.yml (hunk 5 above)
Comment (brb-nv) [reply tests/integration/test_lists/test-db/l0_h100.yml]: Updated!

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 6]
```diff
@@ -341,6 +341,7 @@ def __init__(
                                           None)
         self.guided_decoder = guided_decoder
         self.disable_overlap_scheduler = disable_overlap_scheduler
+        self.enable_early_first_token_emission = enable_early_first_token_emission
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: enable_early_first_token_emission should probably be disabled when return_generation_logits=True. In the early-emission path, the first-token response snapshots logits via get_latest_logits_unexcluded(), but this runs after the next overlap forward step has already appended the next logits chunk. That can make the first streaming response carry logits for the wrong token. Since this feature is an optional TTFT optimization, falling back to the existing response path for requests that return generation logits seems safer unless we add explicit logits parity coverage for streaming=True, return_generation_logits=True, and max_tokens > 1.

File: tensorrt_llm/llmapi/llm_args.py [hunk 7]
```diff
@@ -3857,6 +3816,14 @@ class TorchLlmArgs(BaseLlmArgs):
         "irrespective of confidential compute state.",
         status="prototype")
 
+    enable_early_first_token_emission: bool = Field(
```
Comment (QiJune) [inline tensorrt_llm/llmapi/llm_args.py]: How about renaming this to enable_early_first_token_response? That makes it clearer that this option changes when the first streaming response is sent, not how the first token is generated. emission feels a bit internal/ambiguous for a public LLM API flag.

Comment (brb-nv) [reply tensorrt_llm/llmapi/llm_args.py]: Makes sense. Renamed: https://github.com/NVIDIA/TensorRT-LLM/pull/14061/commits/9de7aac0194cef4c6e4d43733b547fc150fb61f0

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 6 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Great point, Freddy. Adding coverage. I believe this should be supported with current changes:
https://github.com/NVIDIA/TensorRT-LLM/pull/14061/commits/b02a69f5baf2b4b09506852513b89f98be698c80

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 4 above)
Comment (brb-nv) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Discussed with Iman offline that I don't have a good way to test this. Also, I descoped the feature to be opt-in for now so that it doesn't cause unexpected disruption on overlap scheduler codepath.

**Review by QiJune — APPROVED** [review]
LGTM

**Review by yingguo-trt — APPROVED** [review]
